paddle/fluid/framework/ir/generate_pass.h

```diff
@@ -43,6 +42,158 @@ class GeneratePass : public Pass {
+  OpHelper& operator()(std::initializer_list<Arguments> inputs);
+
+  VarHelper Out(const char* name);
+
+ private:
+  OpHelper() = delete;
+  DISABLE_COPY_AND_ASSIGN(OpHelper);
+
+  const char* type_;
+  proto::OpDesc* op_desc_;
+  SubgraphHelper* subgraph_helper_;
+};
+
+/*
+ * SubgraphHelper is used to define pattern/replace subgraphs.
+ *
+ * Use lambda expression to define subgraph like Python. SubgraphHelper
+ * converts lambda expression to ProgramDesc.
+ *
+ * In order to define a subgraph, user need to use VarHelper and OpHelper.
+ * Use the macros instead of class names, so user can develop better and
+ * don't need to know too much about underlying implementation.
+ *
+ * An example of defining a subgraph as follows:
+ *
+ *   SUBGRAPH_(subgraph)([subgraph=&subgraph](VAR_(x), VAR_(y), VAR_(z)) {
+ *     auto ewadd1 = OP_(elementwise_add)({{"X", x}, {"Y", y}}).Out("Out");
+ *     auto ewadd2 = OP_(elementwise_add)({{"X", ewadd1}, {"Y", z}}).Out("Out");
+ *     return ewadd2;
+ *   });
+ *
+ */
+class SubgraphHelper {
+ public:
+  SubgraphHelper() = default;
+  // The lambda expression is a prvalue expression.
+  template <typename T>
+  SubgraphHelper& operator=(const T&& f) {
+    proto::BlockDesc* block = program_desc_.add_blocks();
+    block->set_idx(0);
+    block->set_parent_idx(0);
+    AddOutputVars(f());
+    return *this;
+  }
+
+  proto::ProgramDesc* ProgramDesc();
+  const proto::ProgramDesc& ProgramDesc() const;
+  const std::vector<std::string>& InputVars() const;
+  const std::vector<std::string>& OutputVars() const;
+
+  void AddInputVar(const std::string& name);
+
+  void AddOutputVars(const VarHelper& var_helper);
+
+  template <size_t i, typename... Ts,
+            std::enable_if_t<i + 1 < sizeof...(Ts)>* = nullptr>
+  void AddOutputVars(const std::tuple<Ts...>& outputs) {
+    AddOutputVars(std::get<i>(outputs));
+    AddOutputVars<i + 1>(outputs);
+  }
+
+  template <size_t i, typename... Ts,
+            std::enable_if_t<i + 1 == sizeof...(Ts)>* = nullptr>
+  void AddOutputVars(const std::tuple<Ts...>& outputs) {
+    AddOutputVars(std::get<i>(outputs));
+  }
+
+  template <typename... Ts>
+  void AddOutputVars(const std::tuple<Ts...>& outputs) {
+    AddOutputVars<0>(outputs);
+  }
+
+ private:
+  DISABLE_COPY_AND_ASSIGN(SubgraphHelper);
+  std::vector<std::string> input_vars_;
+  std::vector<std::string> output_vars_;
+  proto::ProgramDesc program_desc_;
+};
+
+}  // namespace generate_pass
+
+class PassPairs {
+ public:
+  using SubgraphType = generate_pass::SubgraphHelper;
+
+  PassPairs() = default;
+  PassPairs(const SubgraphType& pattern, const SubgraphType& replace);
+
+  void AddPassDesc(const SubgraphType& pattern, const SubgraphType& replace);
+
+  const proto::MultiPassDesc& MultiPassDesc() const;
+
+ private:
+  proto::MultiPassDesc multi_pass_desc_;
+};
+
+// Use function to register in CC.
+template <PassPairs (*Functor)(void)>
+class CXXGeneratePass : public GeneratePass {
```

Comment:
Name may not be very good. Maybe "MacroPassHelper" or to help users to know it is used in macro? CXX doesn't have meaning in C++ code.